### DISCUSSION TITLE
NXP PKCS#11 Parsec integration testing.

### DISCUSSION THREAD

**Discussion by sahilnxp** [parallaxsecond/parsec#457] [Q&A] [answered]
Hi,

I have just integrated the Parsec PKCS#11 Provider with NXP PKCS#11 Library and ran already available tests in the PARSEC only and all of them gets passed.
From below dump it looks like there are only 15 tests done, and there is no cryptography tests like key generation/encryption/decryption tests.
I am curious to know if currently only these tests are supported with PKCS#11 or we can do more testing with some other commands ?

Tested this on LS1046 board.

Thanks in advance.
Sahil 

```
root@localhost:~/parsec# git clone https://github.com/tpm2-software/tpm2-tss.git^C
root@localhost:~/parsec# RUST_LOG=info ./target/debug/parsec -c e2e_tests/provider_cfg/pkcs11/config.toml &
[2] 8400
[INFO  parsec] Parsec started. Configuring the service...
[INFO  parsec_service::utils::service_builder] Creating a PKCS 11 Provider.
[INFO  parsec_service::providers::pkcs11] Building a PKCS 11 provider with library '/usr/lib/libpkcs11.so'
[WARN  parsec_service::utils::service_builder] Direct authenticator has been set as the default one. It is only secure under specific requirements. Please make sure to read the Recommendations on a Secure Parsec Deployment at https://parallaxsecond.github.io/parsec-book/parsec_security/secure_deployment.html
[WARN  parsec_service::front::domain_socket] Removing the existing socket file at /tmp/parsec.sock.
[INFO  parsec] Parsec is ready.
root@localhost:~/parsec#
```
```

root@localhost:~/parsec# cargo test --features pkcs11-provider
    Finished test [unoptimized + debuginfo] target(s) in 0.21s
     Running unittests (target/debug/deps/parsec_service-75d8cd3006b96a11)

running 15 tests
test authenticators::unix_peer_credentials_authenticator::test::successful_authentication ... ok
test authenticators::unix_peer_credentials_authenticator::test::unsuccessful_authentication_no_metadata ... ok
test authenticators::unix_peer_credentials_authenticator::test::admin_check ... ok
test authenticators::unix_peer_credentials_authenticator::test::unsuccessful_authentication_wrong_declared_uid ... ok
test authenticators::unix_peer_credentials_authenticator::test::unsuccessful_authentication_wrong_metadata ... ok
test authenticators::unix_peer_credentials_authenticator::test::unsuccessful_authentication_garbage_data ... ok
test key_info_managers::on_disk_manager::test::big_names_emoticons ... ok
test key_info_managers::on_disk_manager::test::big_names_ascii ... ok
test key_info_managers::on_disk_manager::test::exists ... ok
test key_info_managers::on_disk_manager::test::insert_get_key_info ... ok
test key_info_managers::on_disk_manager::test::insert_overwrites ... ok
test providers::core::tests::test_ping ... ok
test key_info_managers::on_disk_manager::test::remove_unexisting_key ... ok
test key_info_managers::on_disk_manager::test::insert_remove_key ... ok
test key_info_managers::on_disk_manager::test::create_and_load ... ok

test result: ok. 15 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests (target/debug/deps/parsec-5665ce8f2f648cee)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests parsec-service

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

root@localhost:~/parsec# 

```

**Comment by hug-dev** [comment, marked as answer]
Hey @sahilnxp 👋!

Good question! Actually what you ran are only the Parsec unit tests. There is indeed a large base of tests that we call the end-to-end tests that are available in the `e2e_tests` folder. Those are crypto and key management tests that are executed against a running Parsec version.

To execute them, have a Parsec service running and then execute:
```
cargo test --features pkcs11-provider --manifest-path ./e2e_tests/Cargo.toml normal_tests
```

Let us know if that does not work, or if you need any help! Hopefully they all pass 🙏 

You can find more information about testing [in the book](https://parallaxsecond.github.io/parsec-book/parsec_service/tests/index.html). We have a list of all of our tests in the subpage.

**Reply by sahilnxp** [reply]
Hi @hug-dev 

I tried running the command you mentioned, lot of tests failed.
Here is the log https://pastebin.com/eGSNzqKH 

All of them failed with `panicked at called Result::unwrap() on an Err value: Client(InvalidSocketAddress)`

Am I doing something wrong ?

**Reply by ionut-arm** [reply]
Hey,

That problem appears because the test client we use doesn't know where to look for the Parsec socket for communication. When the service starts, it creates a socket (normally in [`/run/parsec/parsec.sock`](https://github.com/parallaxsecond/parsec/blob/3dd2ae2122e31c81ede53dd77946d16d804d81e8/src/front/domain_socket.rs#L23) for production environments). However this path can be changed in the configuration file - for example, for our PKCS11 end-to-end tests, we use:
https://github.com/parallaxsecond/parsec/blob/3dd2ae2122e31c81ede53dd77946d16d804d81e8/e2e_tests/provider_cfg/pkcs11/config.toml#L15

I'm guessing you used one of our test configurations, so your socket would be in `/tmp/parsec.sock` as well, in which case you need to set the `PARSEC_SERVICE_ENDPOINT` environment variable before you run the tests:

```bash
export PARSEC_SERVICE_ENDPOINT="unix:/tmp/parsec.sock"
```

@hug-dev maybe we should make the test client automatically connect to `/tmp/parsec.sock`, it is quite annoying to try to run tests locally and hit this problem immediately.

**Reply by sahilnxp** [reply]
Thanks @ionut-arm 
Issue was with accessing the `parsec.sock` only, corrected that some more test cases passed.
Now 65 passed; 46 failed;

Since NXP PKCS#11 is not full PKCS#11 implementation, I think they are expected to fail.

Logs for tests:
https://pastebin.com/Cdff3dqb - PARSEC Service Logs
https://pastebin.com/yBLXtsU8 - PARSEC Client Logs


Thanks 
Sahil

**Reply by ionut-arm** [reply]
Hi,

Those logs don't look too encouraging - the first thing that strikes me is that `create_and_destroy` and `create_and_destroy_ecc` both fail, and those are supposed to be the most basic tests that involve working cryptographic keys. From looking briefly at the list, it seems that most tests that do try to deal with working keys end up failing, mostly by encountering an `PsaErrorCommunicationFailure` response status. From the service logs:
```
[libpkcs11:p11_template_add_attr, 1842] Error: Template type invalid
[libpkcs11:attribute_privk_validate, 1267] Error: Attribute Read Only
...
[ERROR parsec_service::providers::pkcs11::utils] Error converted to PsaErrorCommunicationFailure; Error: An invalid value was specified for a particular attribute in a template.  See Section 4.1 for more information.
```

It sounds like we're trying to set an attribute on the private key that we shouldn't be touching. 

I'm curious, what version of Parsec are you using? A specific commit from our `main` branch, or the latest release - 0.7.x?

**Reply by sahilnxp** [reply]
I am using following latest commit from main branch:


```
commit 94660bd9f4193256e809b8b492d977619e9a1065 (HEAD -> main, origin/main, origin/HEAD)
Merge: 2af44cc 8a86822
Author: Ionu<C8>ihalcea <52527443+ionut-arm@users.noreply.github.com>
Date:   Wed Jun 30 19:38:31 2021 +0100

    Merge pull request #454 from ionut-arm/cross-release

    Make cross-compilation run on release version

commit 8a868229c1ab0a223e79ed9ad73002e53f29f32f
Author: Ionut Mihalcea <ionut.mihalcea@arm.com>
Date:   Mon Jun 28 17:03:03 2021 +0100

    Make cross-compilation run on release version

    This commit changes the command for cross compilation building to work
    on the release version of the binary instead. This is because a recent
    change in the way process liveness is done refers to the release path in
    particular.

    The version of Rust used to run the coverage tests is also changed to
    statically track 1.51.0 to overcome some compilation errors. The SPIFFE
    authenticator is also excluded from the tests, as it was proving
    problematic during compilation.

    Signed-off-by: Ionut Mihalcea <ionut.mihalcea@arm.com>

commit 2af44cc6dc10b23511e9442f4281636728c59a14

```

**Reply by ionut-arm** [reply]
Right! There's one more commit that I merged yesterday and it contains some fixes in the PKCS11 provider - could you try that one as well? If the same errors come up, we'll have to investigate which attributes cause those errors.

**Reply by sahilnxp** [reply]
I just pulled the code and ran again:
Now test result: FAILED. 68 passed; 52 failed; 1 ignored; 0 measured; 29 filtered out; finished in 16.51s

Logs for tests:
PARSEC Service: https://pastebin.com/NHrUqCb0
PARSEC Client: https://pastebin.com/Jhc8wRsr

Please let me know if anything else to try.

**Reply by ionut-arm** [reply]
Ok, not very successful ☹️ Could you please try checking out the `0.7.2` tag/branch (commit `9e7fd2eb6a02c4f9017fdeaa2b6eba2fd73e4b3c`) and running again (see below an alternative command to only run a single test that's a good indicator)? Apologies for the messy approach, unfortunately it seems like the NXP PKCS11 implementation is more strict in some respects and we can't really test it anywhere locally.

From the logs you provided I expect that your implementation rejects one of the attributes we try to set on the private key as read-only. I suspect it's `CKA_EXPORTABLE`, so if you get the same failures you could try commenting out the attributes set on `priv_template` in `fn key_pair_usage_flags_to_pkcs11_attributes` in the `src/providers/pkcs11/utils.rs` file, starting with `priv_template.push(Attribute::Extractable((usage_flags.export).into()));`. Given that it's such a specific problem, you don't need to run all the tests, so you can instead run:
```bash
cargo test --features pkcs11-provider --manifest-path ./e2e_tests/Cargo.toml create_and_destroy
```

**Reply by ionut-arm** [reply]
Ok, it seems we had actually encountered this issue before on Layerscape platforms, and it is indeed due to us setting `CKA_EXPORTABLE` and `CKA_SENSITIVE` on the private keys. I've raised an issue [here](https://github.com/parallaxsecond/parsec/issues/462) for us to fix it. The change shouldn't be too big, but it will only be tagged and released in v0.8.0 in September (most likely).

**Reply by sahilnxp** [reply]
With 0.7.2 tag, getting following error with client:
error: could not compile `picky-asn1-x509`

But not getting this issue with `main `branch

PARSEC Client: https://pastebin.com/xJReg7vB

After removing the CKA_SENSITIVE and CKA_EXPORTABLE attribute: create_and_destroy test passed
create_and_destroy logs: https://pastebin.com/Rf5k0m4K 

After removing the CKA_SENSITIVE and CKA_EXPORTABLE attribute: normal_tests status
`test result: FAILED. 101 passed; 19 failed; 1 ignored; 0 measured; 29 filtered out; finished in 151.96s`
PARSEC Client: https://pastebin.com/eNbhPx3r
PARSEC Service: https://pastebin.com/uMVbSsbi

**Reply by ionut-arm** [reply]
Alright, so it does look like `CKA_SENSITIVE` and `CKA_EXPORTABLE` are the issue - the remaining failing tests are trying to do public key operations like doing asymmetric encryption and signature verification, which as far as I know aren't supported on the NXP board.

You can fix those by setting the following flag on in your `config.toml`, in the PKCS11 `[[provider]]` section:
```toml
software_public_operations = true
```

However, for this to work you do need the Mbed Crypto library installed locally. If you want to do that you can follow easy steps [here](https://github.com/parallaxsecond/parsec/blob/3dd2ae2122e31c81ede53dd77946d16d804d81e8/e2e_tests/docker_image/parsec-service-test-all.Dockerfile#L71).

**Reply by hug-dev** [reply]
> However, for this to work you do need the Mbed Crypto library installed locally. If you want to do that you can follow easy steps here.

You can try to ignore that at first, I believe the build system will build Mbed Crypto from scratch.

**Reply by sahilnxp** [reply]
It gets more better now after enabling 
`software_public_operations = true`

Status is:
test result: FAILED. 112 passed; 8 failed; 1 ignored; 0 measured; 29 filtered out; finished in 152.41s

PARSEC Client: https://pastebin.com/yjqhgK2w
PARSEC Service: https://pastebin.com/jJYYnRQA

@hug-dev Installed the mbedtls before seeing your message. :-)

**Reply by hug-dev** [reply]
Nice, going in the right direction 👌 

I had a look at the failed tests, I think it would be worth to investigate further.

- 5 of them seem to fail on `verify_with_ecdsa_sha256`. It would be worth to execute only one of them (say `simple_verify_hash_ecc`) and debug what's wrong. Maybe the ECDSA mechanism is not supported?
- two of them (`verify_ecc_with_ring` and `export_ecc_public_key`) fail when exporting the public part of a generated ECC key. Might be worth to debug one of them.
- the format of a public key is wrong in `check_public_ecc_export_format2`

**Reply by hug-dev** [reply]
> Maybe the ECDSA mechanism is not supported?

That seems to be the case from the list of supported mechanisms [here](https://docs.nxp.com/bundle/GUID-3B3DA6F9-635B-4E21-AA0D-439A9ACF044B/page/GUID-507B777E-B29B-44E9-83E3-2FDF5691DC2E.html). Only RSA ones seem to be supported 😢 
When capability profiling is available, we should maybe use `C_GetMechanismList` to dynamically check which ones are supported.
Might still be a bug of us not returning `PsaErrorNotSupported` when those are used?

**Reply by ionut-arm** [reply]
Yes, that sounds like an excellent idea, and we can/should do the same for TPMs. I guess we've not had a clear "requirement" for such discovery until now.

**Reply by hug-dev** [reply]
Ok I went for some detective work 🕵️ 

When verifying, because the PKCS11 library does not support it, we export the public key to the Mbed Crypto provider and do the verification there. This is done through the [`move_pub_key_to_psa_crypto`](https://github.com/parallaxsecond/parsec/blob/e244c7d8e3e7c60a6a9499735a395af3bb3f6463/src/providers/pkcs11/key_management.rs#L50) call.
I counted the `Attempting to export public key` and the `Importing public key into PSA Crypto` occurences in your logs and there is 8 more (!!) of the first one which maps to the eight tests failing!!

If that's the case, I think the `PsaErrorGenericError` returned is from [that line](https://github.com/parallaxsecond/parsec/blob/e244c7d8e3e7c60a6a9499735a395af3bb3f6463/src/providers/pkcs11/key_management.rs#L402), because the `EC_POINT` attribute returned by the PKCS11 library is not correctly parsed as an ASN.1 OctetString.

It could be interesting to expend that error log message to print the content of that attribute to see what is its form?

**Reply by ionut-arm** [reply]
Btw, note for future - when looking at failing tests, adding `-- --test-threads=1` would be useful, so that each test is run sequentially and you don't have to sift through logs from different threads overlapping.

**Comment by hug-dev** [comment]
Also, feel free to join [our Slack channel](https://github.com/parallaxsecond/community#community-channel) if you want and if that helps to debug this!

**Comment by hug-dev** [comment]
I am wondering if the documentation of the [NXP PKCS11](https://docs.nxp.com/bundle/GUID-3B3DA6F9-635B-4E21-AA0D-439A9ACF044B/page/GUID-507B777E-B29B-44E9-83E3-2FDF5691DC2E.html) library is up-to-date.

We are using way more PKCS11 calls than what is listed there. Also, all of our sessions are R/W User sessions without it seeming to cause any problems? We are also definitely generating and importing a lot of keys using the PKCS11 library.

**Reply by sahilnxp** [reply]
Yes, documentation is not updated.
Currently supported mechanisms are:
```
CKM_MD5
CKM_SHA_1
CKM_SHA256
CKM_SHA384
CKM_SHA512
CKM_RSA_PKCS
CKM_MD5_RSA_PKCS
CKM_SHA1_RSA_PKCS
CKM_SHA256_RSA_PKCS
CKM_SHA384_RSA_PKCS
CKM_SHA512_RSA_PKCS
CKM_ECDSA_SHA1
CKM_ECDSA
CKM_RSA_PKCS_KEY_PAIR_GEN
CKM_EC_KEY_PAIR_GEN
CKM_RSA_PKCS_OAEP
```

If you want to have a look at PKCS#11 implementation you can have at:
https://source.codeaurora.org/external/qoriq/qoriq-components/libpkcs11/

**Reply by hug-dev** [reply]
Nice, thanks for the link! We also found more up-to-date documentation [here](https://docs.nxp.com/bundle/GUID-3FFCCD77-5220-414D-8664-09E6FB1B02C6/page/GUID-507B777E-B29B-44E9-83E3-2FDF5691DC2E.html) 👌

**Comment by hug-dev** [comment]
Hi @sahilnxp 👋 !

Just to update you that https://github.com/parallaxsecond/parsec/pull/466 was merged and provide a better way to handle one of the problems that you saw. You should be able to use Parsec from `main` and set the `allow_export` flag to `false`.
There might be other issues as reported [above](https://github.com/parallaxsecond/parsec/discussions/457#discussioncomment-979310), maybe related with the format used to represent keys internally. Happy to help if you want to debug it more.

We were also curious and interested about your motivation for this work? We would be definitely excited about greater integration between Parsec and the Layerscape 👌 

It might be interesting for you to check out this issue (https://github.com/parallaxsecond/parsec/issues/439) which is about tighter integration with the `libsecure_obj` library directly.
